README.md
<div>
    <a href="https://travis-ci.org/pjfl/p5-file-unixauth"><img src="https://travis-ci.org/pjfl/p5-file-unixauth.svg?branch=master" alt="Travis CI Badge"></a>
    <a href="http://badge.fury.io/pl/File-UnixAuth"><img src="https://badge.fury.io/pl/File-UnixAuth.svg" alt="CPAN Badge"></a>
    <a href="http://cpants.cpanauthors.org/dist/File-UnixAuth"><img src="http://cpants.cpanauthors.org/dist/File-UnixAuth.png" alt="Kwalitee Badge"></a>
</div>

# Name

File::UnixAuth - Read and write the Unix authentication files

# Version

Describes version v0.26.$Rev: 3 $ of [File::UnixAuth](https://metacpan.org/pod/File::UnixAuth)

# Synopsis

    use File::UnixAuth;

    my $unixauth_ref = File::UnixAuth->new( $unixauth_attributes );

# Description

Extends [File::DataClass::Schema](https://metacpan.org/pod/File::DataClass::Schema). Provides for the reading and
writing of the the Unix `/etc/group`, `/etc/passwd`, and
`/etc/shadow` files.

Since these files share a common format they all use the the same
storage class [File::UnixAuth::Storage](https://metacpan.org/pod/File::UnixAuth::Storage). Defines three result
sources; `group`, `passwd`, and `shadow`

# Configuration and Environment

Defines these attributes;

- `post_update_hook`

    A code reference that is called after an update completes. Defaults to
    `undef`. If set to:

        sub { qx( 'grpconv' ) }

    then the file `/etc/gshadow` will be updated

- `result_source_attributes`

    Defines the result sources and their attributes

- `source_name`

    A required string. Selects the required result source. Set to one of;
    `group`, `passwd`, or `shadow`

- `storage_attributes`

    Change the defaults to create a backup file with a `.bak` extension

Modifies these methods;

- `resultset`
- `source`

# Subroutines/Methods

None

# Diagnostics

None

# Dependencies

- [File::DataClass::Schema](https://metacpan.org/pod/File::DataClass::Schema)
- [File::UnixAuth::Result](https://metacpan.org/pod/File::UnixAuth::Result)
- [Moo](https://metacpan.org/pod/Moo)

# Incompatibilities

There are no known incompatibilities in this module

# Bugs and Limitations

There are no known bugs in this module.
Please report problems to the address below.
Patches are welcome

# Author

Peter Flanigan, `<pjfl@cpan.org>`

# License and Copyright

Copyright (c) 2021 Peter Flanigan. All rights reserved

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself. See [perlartistic](https://metacpan.org/pod/perlartistic)

This program is distributed in the hope that it will be useful,
but WITHOUT WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE


Changes
Revision history for File::UnixAuth

          - Refactored coding style
          - Toolchain update

0.26.1    2015-11-13 00:34:52
          - Another broken smoker added to exception list

0.25.1    2015-11-07 14:01:52
          - Strictified POD encoding
          - Travis uses containerised builds
          - Toolchain update

0.24.1    2015-05-31 14:24:43
          - Dropped Lingua::EN::NameParse
            There is something quite wrong with that package

0.23.1    2015-05-31 13:02:28
          - Added badges and public repository
          - Updated copyright year
          - Updated prereqs
          - Storage API methods now called read_from_file and write_to_file
          - Dropped Class::Method::ModifiersX::Augment

0.22.1    2014-07-05 10:18:56
          - Replaced namespace::sweep with namespace::autoclean
          - ::Result::_source attr name change to result_source

0.21.1    2014-05-28 11:29:58
          - Added post_update_hook callback to support /etc/gshadow
            Closes RT#96026

0.20.1    2014-05-27 19:29:19
          - Removed Ident and VERSION lines
          - Updated toolchain

0.19.1    2013-08-15 19:39:58
          - Switched to Dist::Zilla

0.18.3    2013-08-07 18:23:12
          - Reset min perl testing to original behaviour
          - Raised minimum Perl version to v5.10.1
          - Updated POD


t/boilerplate.pm
package t::boilerplate;

use strict;
use warnings;
use File::Spec::Functions qw( catdir updir );
use FindBin               qw( $Bin );
use lib               catdir( $Bin, updir, 'lib' ), catdir( $Bin, 'lib' );

use Module::Build;
use Sys::Hostname;

sub plan (;@) {
   $_[ 0 ] eq 'skip_all' and print '1..0 # SKIP '.$_[ 1 ]."\n" and exit 0;
}

my ($builder, $host, $notes, $perl_ver);

BEGIN {
   $host     = lc hostname;
   $builder  = eval { Module::Build->current };
   $notes    = $builder ? $builder->notes : {};
   $perl_ver = $notes->{min_perl_version} || 5.008;

   eval { require Test::Requires }; $@ and plan skip_all => 'No Test::Requires';

   $Bin =~ m{ : .+ : }mx and plan skip_all => 'Two colons in $Bin path';

   if ($notes->{testing}) {
      $host eq 'foobar.spo.virtua.com.br' and plan skip_all =>
         'Broken smoker 35c6ef92-842a-11e5-8e8c-72a0c69edca9';
      $host eq 'vm-debian-001' and plan skip_all =>
         'Broken smoker da099c60-896d-11e5-b552-e159351a082c';
   }
}

use Test::Requires "${perl_ver}";
use Test::Requires { version => 0.88 };

sub import {
   strict->import;
   $] < 5.008 ? warnings->import : warnings->import( NONFATAL => 'all' );
   return;
}

1;

# Local Variables:
# mode: perl
# tab-width: 3
# End:
# vim: expandtab shiftwidth=3:


Build.PL
use strict;
use warnings;

use Module::Build 0.4004;

my $perl_ver = 5.010001;
my $name     = 'File-UnixAuth'; $name =~ s{ - }{::}gmx;
my $prereqs  = {
  'build_requires' => {
    'Module::Build' => '0.4004',
    'version' => '0.88'
  },
  'configure_requires' => {
    'Module::Build' => '0.4004',
    'version' => '0.88'
  },
  'requires' => {
    'File::DataClass' => 'v0.66.0',
    'Moo' => '2.000001',
    'namespace::autoclean' => '0.26',
    'perl' => '5.010001'
  },
  'test_requires' => {
    'File::Spec' => '0',
    'Hash::MoreUtils' => '0.05',
    'Module::Build' => '0.4004',
    'Module::Metadata' => '0',
    'Sys::Hostname' => '0',
    'Text::Diff' => '0',
    'version' => '0.88'
  },
};
my $testing  = sub { $ENV{AUTOMATED_TESTING}
                  || $ENV{PERL_CPAN_REPORTER_CONFIG}
                  || $ENV{PERL5_CPANPLUS_IS_VERSION} };
my $whimper  = sub { print {*STDOUT} $_[ 0 ]."\n"; exit 0 };
my $stop     = sub {
   $testing->() or $whimper->( $_[ 0 ] );
   $prereqs->{build_requires} = { %{ $prereqs->{configure_requires} } };
   $prereqs->{test_requires } = { %{ $prereqs->{configure_requires} } };
   $prereqs->{test_requires }->{ 'Test::Requires' } = 0.06;
   delete $prereqs->{requires};
};

# Not interested in reports below min perl ver
$] < $perl_ver and $stop->( "Minimum required Perl version is ${perl_ver}" );

Module::Build->new(
   'module_name' => $name,
   'notes'       => { min_perl_version => $perl_ver, testing => $testing->() },
   %{ $prereqs },

)->create_build_script;

exit 0;


lib/File/UnixAuth/Result.pm
package File::UnixAuth::Result;

use namespace::autoclean;

use File::DataClass::Constants qw( FALSE );
use File::DataClass::Functions qw( is_member );
use Moo;

extends q(File::DataClass::Result);

after 'update' => sub {
   my $self   = shift;
   my $source = $self->can('result_source')
              ? $self->result_source : $self->_source;
   my $hook   = $source->schema->post_update_hook;

   $hook->($self) if $hook;

   return;
};

sub add_user_to_group {
   my ($self, $user) = @_;

   my $users = $self->members;

   return FALSE if is_member $user, $users;

   $self->members([ @{$users}, $user ]);

   return $self->update;
}

sub remove_user_from_group {
   my ($self, $user) = @_;

   my $users = $self->members;

   return FALSE unless is_member $user, $users;

   $self->members([ grep { $_ ne $user } @{$users} ]);

   return $self->update;
}

1;

__END__

=pod

=encoding utf-8

=head1 Name

File::UnixAuth::Result - Unix authentication and authorisation file custom results

=head1 Synopsis

   use File::UnixAuth::Result;

=head1 Description

=head1 Configuration and Environment

=head1 Subroutines/Methods

=head2 C<update>

Modifies the C<update> method call. Calls the post update hook if defined

=head2 add_user_to_group

Adds a user to a group

=head2 remove_user_from_group

Removes a user from a group

=head1 Diagnostics

None

=head1 Dependencies

=over 3

=item L<File::DataClass::Result>

=back

=head1 Incompatibilities

There are no known incompatibilities in this module

=head1 Bugs and Limitations

There are no known bugs in this module.
Please report problems to the address below.
Patches are welcome

=head1 Author

Peter Flanigan, C<< <Support at RoxSoft.co.uk> >>

=head1 License and Copyright

Copyright (c) 2021 Peter Flanigan. All rights reserved

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself. See L<perlartistic>

This program is distributed in the hope that it will be useful,
but WITHOUT WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE

=cut

# Local Variables:
# mode: perl
# tab-width: 3
# End:


lib/File/UnixAuth/Storage.pm
package File::UnixAuth::Storage;

use namespace::autoclean;

use File::DataClass::Constants qw( FALSE NUL SPC TRUE );
use Moo;

extends q(File::DataClass::Storage);

# Public methods
sub read_from_file {
   my ($self, $rdr) = @_;

   $self->encoding and $rdr->encoding($self->encoding);

   return $self->_read_filter([$rdr->chomp->getlines]);
}

sub write_to_file {
   my ($self, $wtr, $data) = @_;

   $wtr->encoding($self->encoding) if $self->encoding;

   $wtr->println(@{$self->_write_filter($data)});

   return $data;
}

# Private methods
sub _deflate {
   my ($self, $hash, $id) = @_;

   my $attr  = $hash->{$id};
   my $gecos = NUL;

   $attr->{members} = join ',', @{$attr->{members} || []}
      if exists $attr->{members};

   $gecos .= $attr->{first_name} // NUL if exists $attr->{first_name};

   $gecos .= $attr->{last_name} ? SPC.$attr->{last_name} : NUL
      if exists $attr->{last_name};

   if ($attr->{location} || $attr->{work_phone} || $attr->{home_phone}) {
      $gecos .= ','.($attr->{location  } // NUL);
      $gecos .= ','.($attr->{work_phone} // NUL);
      $gecos .= ','.($attr->{home_phone} // NUL);
   }

   $attr->{gecos} = $gecos if $gecos;

   return;
}

sub _inflate {
   my ($self, $hash, $id) = @_;

   my $attr = $hash->{$id};

   $attr->{members} = [ split m{ , }mx, $attr->{members} // NUL ]
      if exists $attr->{members};

   if (exists $attr->{gecos}) {
      my @fields = qw( full_name location work_phone home_phone );

      @{$attr}{@fields} = split m{ , }mx, $attr->{gecos} // NUL;

      my $names = _parse_name($attr->{full_name});

      $attr->{first_name} = $names->{first_name} // NUL;
      $attr->{last_name } = $names->{last_name } // NUL;
      delete $attr->{full_name};
      delete $attr->{gecos};
   }

   return;
}

sub _read_filter {
   my ($self, $buf) = @_;

   my $hash        = {};
   my $order       = 0;
   my $source_name = $self->schema->source_name;
   my $fields      = $self->schema->source->attributes;

   for my $line (@{$buf || []}) {
      my ($id, @rest) = split m{ : }mx, $line;
      my %attr = ();

      @attr{@{$fields}} = @rest;
      $attr{_order_by} = $order++;
      $hash->{$id} = \%attr;
      $self->_inflate($hash, $id);
   }

   return { $source_name => $hash };
}

sub _write_filter {
   my ($self, $data) = @_;

   my $buf         = [];
   my $source_name = $self->schema->source_name;
   my $fields      = $self->schema->source->attributes;
   my $hash        = $data->{$source_name};

   $fields = [ @{$fields}[0 .. 5] ] if $source_name eq 'passwd';

   for my $id (sort { _original_order($hash, $a, $b) } keys %{$hash}) {
      $self->_deflate($hash, $id);

      my $attr = $hash->{$id};

      delete $attr->{_order_by};

      my $line = join ':', map { $attr->{$_} // NUL } @{$fields};

      push @{$buf}, "${id}:${line}";
   }

   return $buf;
}

# Private functions
sub _original_order {
   my ($hash, $lhs, $rhs) = @_;

   # New elements will be  added at the end
   return  1 unless exists $hash->{$lhs}->{_order_by};
   return -1 unless exists $hash->{$rhs}->{_order_by};
   return $hash->{$lhs}->{_order_by} <=> $hash->{$rhs}->{_order_by};
}

sub _parse_name {
   my $full_name = shift;

   return {} unless $full_name;

   my ($first_name, $last_name) = split SPC, $full_name, 2;

   return { first_name => $first_name, last_name => $last_name };
}

1;

__END__

=pod

=encoding utf-8

=head1 Name

File::UnixAuth::Storage - Unix authentication and authorisation file storage

=head1 Synopsis

   use File::UnixAuth::Storage;

=head1 Description

=head1 Configuration and Environment

=head1 Subroutines/Methods

=head2 read_from_file

=head2 write_to_file

=head1 Diagnostics

=head1 Dependencies

=over 3

=item L<File::DataClass::Storage>

=item L<MooX::Augment>

=back

=head1 Incompatibilities

There are no known incompatibilities in this module

=head1 Bugs and Limitations

There are no known bugs in this module.
Please report problems to the address below.
Patches are welcome

=head1 Author

Peter Flanigan, C<< <Support at RoxSoft.co.uk> >>

=head1 License and Copyright

Copyright (c) 2021 Peter Flanigan. All rights reserved

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself. See L<perlartistic>

This program is distributed in the hope that it will be useful,
but WITHOUT WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE

=cut

# Local Variables:
# mode: perl
# tab-width: 3
# End:


lib/File/UnixAuth.pm
package File::UnixAuth;

use 5.010001;
use namespace::autoclean;
use version; our $VERSION = qv( sprintf '0.26.%d', q$Rev: 3 $ =~ /\d+/gmx );

use File::DataClass::Constants qw( NUL TRUE );
use File::DataClass::Types     qw( CodeRef Maybe Str );
use File::UnixAuth::Result;
use Moo;

extends q(File::DataClass::Schema);

has 'post_update_hook' => is => 'ro', isa => Maybe[CodeRef];

has '+result_source_attributes' =>
   default => sub {
      return {
         group               => {
            attributes       => [ qw( password gid members ) ],
            defaults         => { password => 'x' },
            resultset_attributes => {
               result_class  => 'File::UnixAuth::Result',
            },
         },
         passwd              => {
            attributes       => [
               qw( password id pgid gecos homedir shell first_name last_name
                   location work_phone home_phone )
            ],
            defaults         => { password => 'x' },
         },
         shadow              => {
            attributes       => [
               qw( password pwlast pwnext pwafter pwwarn pwexpires pwdisable
                   reserved )
            ],
            defaults         => {
               password      => '*',
               pwlast        => 0,
               pwnext        => 0,
               pwafter       => 99_999,
               pwwarn        => 7,
               pwexpires     => 90,
               reserved      => NUL,
            },
         },
      };
   };

has '+storage_attributes' => default => sub { { backup => '.bak', } };

has '+storage_class' => default => '+File::UnixAuth::Storage';

has 'source_name' => is => 'ro', isa => Str, required => TRUE;

around 'source' => sub {
   my ($orig, $self) = @_;

   return $self->$orig($self->source_name);
};

around 'resultset' => sub {
   my ($orig, $self) = @_;

   return $self->$orig($self->source_name);
};

1;

__END__

=pod

=encoding utf-8

=begin html

<a href="https://travis-ci.org/pjfl/p5-file-unixauth"><img src="https://travis-ci.org/pjfl/p5-file-unixauth.svg?branch=master" alt="Travis CI Badge"></a>
<a href="http://badge.fury.io/pl/File-UnixAuth"><img src="https://badge.fury.io/pl/File-UnixAuth.svg" alt="CPAN Badge"></a>
<a href="http://cpants.cpanauthors.org/dist/File-UnixAuth"><img src="http://cpants.cpanauthors.org/dist/File-UnixAuth.png" alt="Kwalitee Badge"></a>

=end html

=head1 Name

File::UnixAuth - Read and write the Unix authentication files

=head1 Version

Describes version v0.26.$Rev: 3 $ of L<File::UnixAuth>

=head1 Synopsis

   use File::UnixAuth;

   my $unixauth_ref = File::UnixAuth->new( $unixauth_attributes );

=head1 Description

Extends L<File::DataClass::Schema>. Provides for the reading and
writing of the the Unix F</etc/group>, F</etc/passwd>, and
F</etc/shadow> files.

Since these files share a common format they all use the the same
storage class L<File::UnixAuth::Storage>. Defines three result
sources; C<group>, C<passwd>, and C<shadow>

=head1 Configuration and Environment

Defines these attributes;

=over 3

=item C<post_update_hook>

A code reference that is called after an update completes. Defaults to
C<undef>. If set to:

   sub { qx( 'grpconv' ) }

then the file F</etc/gshadow> will be updated

=item C<result_source_attributes>

Defines the result sources and their attributes

=item C<source_name>

A required string. Selects the required result source. Set to one of;
C<group>, C<passwd>, or C<shadow>

=item C<storage_attributes>

Change the defaults to create a backup file with a F<.bak> extension

=back

Modifies these methods;

=over 3

=item C<resultset>

=item C<source>

=back

=head1 Subroutines/Methods

None

=head1 Diagnostics

None

=head1 Dependencies

=over 3

=item L<File::DataClass::Schema>

=item L<File::UnixAuth::Result>

=item L<Moo>

=back

=head1 Incompatibilities

There are no known incompatibilities in this module

=head1 Bugs and Limitations

There are no known bugs in this module.
Please report problems to the address below.
Patches are welcome

=head1 Author

Peter Flanigan, C<< <pjfl@cpan.org> >>

=head1 License and Copyright

Copyright (c) 2021 Peter Flanigan. All rights reserved

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself. See L<perlartistic>

This program is distributed in the hope that it will be useful,
but WITHOUT WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE

=cut

# Local Variables:
# mode: perl
# tab-width: 3
# End:
